contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,201 @@
+  directives is a more complicated approach than modifying the object read from the server
+  and updating it with a PUT.
+
+#### New approach
+
+- Read the live object
+- Compare the live object to last-recorded and local files
+- Update the fields on the live object that was read
+- Send a PUT to update the modified object
+- If encountering optimistic lock failure, retry after re-reading the object
+
+#### Trade offs
+
+Good:
+
+- Simpler code to maintain
+- Behaves correctly in more instances
+- Easy to look at the output and determine the final state
+
+Bad:
+
+- Larger request sizes to the server than when sending a patch
+
+### Restructure code into modular components
+
+In the current implementation of apply - parsing and traversing the object trees, diffing the
+contents and generating the patch are logically entangled.  This creates maintenance and
+testing challenges.  We should instead modularize discrete responsibilities - collating the
+object values and updating the target object.
+
+#### Phase 1: Parse last-recorded, local, live objects and collate
+
+Provide a structure that contains the last, local and live value for each field.  This
+will make it easy to walk the a single tree when making decisions about how to update the object.
+
+#### Phase 2: Diff and update objects
+
+Use the visitor pattern to encapsulate how to update each field type for each merge strategy.
+Unit test each visit function.
+
+## Alternatives considered
+
+Continue to use patch for apply.  Continue to evolve the PATCH semantics to be more
+forgiving of version skewed environments.  Add new directives as requirements identified.
+
+## Detailed analysis of structure and impact today
+
+The following PRs constitute the focus of ~6 months of engineering work.  Each of the PRs is very complex
+for the work what it is solving.
+
+### Patterns observed
+
+- PRs frequently closed or deferred because maintainers / reviewers cannot reason about the impact or
+  correctness of the changes
+- Relatively simple changes
+  - are 200+ lines of code
+  - modify dozens of existing locations in the code
+  - are spread across 1000+ lines of existing code
+- Changes that add new directives require updates in multiple locations - create patch + apply patch
+
+### PRs
+
+[38665](https://github.com/kubernetes/kubernetes/pull/38665/files)
+- Support deletion of primitives from lists
+- Lines (non-test): ~200
+- ~6 weeks
+[44597](https://github.com/kubernetes/kubernetes/pull/44597/files)
+- Support deleting fields not listed in the patch
+- Lines (non-test): ~250
+- ~6 weeks
+[45980](https://github.com/kubernetes/kubernetes/pull/45980/files#diff-101008d96c4444a5813f7cb6b54aaff6)
+- Keep ordering of items when merging lists
+- Lines (non-test): ~650
+[46161](https://github.com/kubernetes/kubernetes/pull/46161/files#diff-101008d96c4444a5813f7cb6b54aaff6)
+- Support using multiple fields for a merge key
```

Comment:
This seems like it will still be a problem.

Reply:
Maybe we can introduce multi-fields merge key as a new tag for apply. It may be the right chance to fix the wrong merge key issue, e.g. container.ports.
Server-side apply in the future can reuse this new merge key.

Reply:
It will still be a problem for PATCH, but provides a simpler solution for apply.

We won't need to come up with a strategy for communicating the merge keys within the PATCH (either explicitly or implicitly through versioning).  A relatively simple solution would be communicate the updated merge-keys to the client in openAPI, and then the client can decide if they want to use these keys or not.

Solving this for PATCH is a larger issue.

Reply:
Patches shouldn't need to communicate merge keys. That should be a property of the schema, not a property of the patch.

Reply:
If it is not explicitly part of the patch, then we need to introduce versioning semantics so both the client and server agree on the implicit merge keys since they may change.